net/sched/sch_gkprio.c

```diff
@@ -0,0 +1,309 @@
+{
+	int prio;
+	
+	for (prio = q->highest_prio - 1; prio >= q->lowest_prio; prio--) {	
+		if (q->qdiscs[prio].qlen > 0)
+			return prio;
+	}
+
+	/* GK queue is empty, return 0 (default highest priority setting). */
+	return 0;
+}
+
+static u16 calc_new_low_prio(const struct gkprio_sched_data *q)
+{
+	int prio;
+
+	for (prio = q->lowest_prio + 1; prio <= q->highest_prio; prio++) {
+		if (q->qdiscs[prio].qlen > 0)
+			return prio;
+	}
+
+	/* GK queue is empty, return GKPRIO_MAX_PRIORITY - 1
+	 * (default lowest priority setting).
+	 */
+	return GKPRIO_MAX_PRIORITY - 1;
+}
+
+static int gkprio_enqueue(struct sk_buff *skb, struct Qdisc *sch,
+			   struct sk_buff **to_free)
+{
+	struct sk_buff_head *qdisc;
+	struct sk_buff_head *lp_qdisc;
+	struct sk_buff *to_drop;
+	int wlen;
+	u16 prio, lp;
+	struct gkprio_sched_data *q;
+
+	/* Obtain the priority of @skb. */
+	wlen = skb_network_offset(skb);
+	switch (tc_skb_protocol(skb)) {
+	case htons(ETH_P_IP):
+		wlen += sizeof(struct iphdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+		prio = ipv4_get_dsfield(ip_hdr(skb)) >> 2;
+		break;
+
+	case htons(ETH_P_IPV6):
+		wlen += sizeof(struct ipv6hdr);
+		if (!pskb_may_pull(skb, wlen))
+			goto drop;
+		prio = ipv6_get_dsfield(ipv6_hdr(skb)) >> 2;
+		break;
+
+	default:
+		prio = q->noip_dfltp;
+		break;
+	}
+
+	q = qdisc_priv(sch);
+	qdisc = &q->qdiscs[prio];
+
+	if (sch->q.qlen < q->max_limit) {
+		skb_queue_tail(qdisc, skb);
+		qdisc_qstats_backlog_inc(sch, skb);
+
+		/* Check to update highest and lowest priorities. */
+		if (prio > q->highest_prio)
+			q->highest_prio = prio;
+
+		if (prio < q->lowest_prio)
+			q->lowest_prio = prio;
+
+		sch->q.qlen++;
+		return NET_XMIT_SUCCESS;
+	}
+
+	/* If this packet is lowest priority, drop it. */
+	lp = q->lowest_prio;
+	if (prio <= lp)
+		return qdisc_drop(skb, sch, to_free);
+
+	/* Drop the packet at the tail of lowest priority qdisc. */
+	lp_qdisc = &q->qdiscs[lp];
+	to_drop = lp_qdisc->prev;
+
+        if (likely(to_drop)) {
+                lp_qdisc->prev = to_drop->prev;
+                lp_qdisc->qlen--;
+                if (!lp_qdisc->prev)
+                        lp_qdisc->next = NULL;
+                to_drop->prev = NULL;
+		qdisc_qstats_backlog_dec(sch, to_drop);
+		 __qdisc_drop(to_drop, to_free);
+		qdisc_qstats_drop(sch);
+        }	
+	else goto drop;
+
+	skb_queue_tail(qdisc, skb);
+	qdisc_qstats_backlog_inc(sch, skb);
+
+	/* Check to update highest and lowest priorities. */
+	if (lp_qdisc->qlen == 0)
+		q->lowest_prio = calc_new_low_prio(q);
+
+	if (prio > q->highest_prio)
+		q->highest_prio = prio;
+
+	return NET_XMIT_SUCCESS;
+drop:
+	qdisc_drop(skb, sch, to_free);
+	return NET_XMIT_SUCCESS | __NET_XMIT_BYPASS;
+}
+
+static struct sk_buff *gkprio_dequeue(struct Qdisc *sch)
+{
+	struct gkprio_sched_data *q = qdisc_priv(sch);
+	struct sk_buff *skb;
+	struct sk_buff_head *hpq = &q->qdiscs[q->highest_prio];
+
+	skb = skb_dequeue(hpq);
```

Comment:
Style: (1) add the line `struct sk_buff *skb = skb_dequeue(hpq);` after line `struct sk_buff_head *hpq = &q->qdiscs[q->highest_prio];`, and (2) drop the current lines: `struct sk_buff *skb;` and `skb = skb_dequeue(hpq);`.

Can't we replace `skb_dequeue()` with `__skb_dequeue()`? Does anyone scheduler aquire lock(s) while it enqueues and dequeues socket buffers to its datastructure?

Reply:
Yes, will make the change. Since the api for schedulers automatically locks the scheduler for qdisc ops, if we'd need to change variables/auxillary datastructures outside of the qdisc  itself, then we'd have to accquire the existing lock appropriately. Accessing the qdisc is taken care off by existing kernel code, none of the current scheduler worries about that while enqueueing or dequeueing.